lib/ood_core/job/adapters/fork/launcher.rb

```diff
@@ -0,0 +1,241 @@
+  end
+
+  def list_remote_sessions(host: nil)
+    host_list = (host) ? [host] : ssh_hosts
+
+    host_list.map {
+      |hostname| list_remote_tmux_session(hostname)
+    }.flatten.sort_by {
+      |hsh| hsh[:session_name]
+    }
+  end
+
+  private
+
+  # Call a forked Slurm command for a given cluster
+  def call(cmd, *args, env: {}, stdin: "")
+    args  = args.map(&:to_s)
+    env = env.to_h
+    o, e, s = Open3.capture3(env, cmd, *args, stdin_data: stdin.to_s)
+    s.success? ? o : raise(Error, e)
+  end
+
+  # The SSH invocation to send a command
+  # -t Force pseudo-terminal allocation (required to allow tmux to run)
+  # -o BatchMode=yes (set mode to be non-interactive)
+  # if ! strict_host_checking
+  # -o UserKnownHostsFile=/dev/null (do not update the user's known hosts file)
+  # -o StrictHostKeyChecking=no (do no check the user's known hosts file)
+  def ssh_cmd(destination_host)
+    if strict_host_checking
+      [
+        'ssh', '-t',
+        '-o', 'BatchMode=yes',
+        "#{username}@#{destination_host}"
+      ]
+    else
+      [
+        'ssh', '-t',
+        '-o', 'BatchMode=yes',
+        '-o', 'UserKnownHostsFile=/dev/null',
+        '-o', 'StrictHostKeyChecking=no',
+        "#{username}@#{destination_host}"
+      ]
+    end
+  end
+
+  # Wraps a user-provided script into a Tmux invocation
+  def wrapped_script(script, session_name)
+    ERB.new(
+      File.read(Pathname.new(__dir__).join('templates/script_wrapper.erb.sh'))
+    ).result(binding.tap {|bnd|
+      {
+        'arguments' => script_arguments(script),
+        'cd_to_workdir' => (script.workdir) ? "cd #{script.workdir}" : '',
+        'debug' => debug,
+        'email_on_terminated' => script_email_on_event(script, 'terminated'),
+        'email_on_start' => script_email_on_event(script, 'started'),
+        'environment' => export_env(script.job_environment),
+        'error_path' => (script.error_path) ? script.error_path.to_s : '/dev/null',
+        'job_name' => script.job_name.to_s,
+        'output_path' => (script.output_path) ? script.output_path.to_s : '/dev/null',
+        'script_content' => script.content,
+        'script_timeout' => script_timeout(script),
+        'session_name' => session_name,
+        'singularity_bin' => singularity_bin,
+        'singularity_image' => singularity_image(script),
+        'tmux_bin' => tmux_bin,
+      }.each{
+        |key, value| bnd.local_variable_set(key, value)
+      }
+    })
+  end
+
+  # Generate the environment export block for this script
+  def export_env(environment)
+    (environment ? environment : {}).tap{
+      |hsh|
+      hsh['SINGULARITY_BINDPATH'] = singularity_bindpath(environment)
+      hsh.delete('SINGULARITY_CONTAINER')
+    }.map{
+      |key, value| "export #{key}=#{Shellwords.escape(value)}"
+    }.sort.join("\n")
+  end
+
+  def singularity_image(script)
+    image = script.job_environment['SINGULARITY_CONTAINER']
+    (image) ? image : default_singularity_image
+  end
```

Comment:
The environment variable `SINGULARITY_CONTAINER` is set by Singularity at run time to point to the image, but is not usually used to define the image-to-be-run as it is here. The alternative to using the environment would be to use the `native` property on the `OodCore::Job::Script`. I chose to use the environment because we are already using the environment to set `SINGULARITY_BINDPATH`, and because in general we discourage the use of `native`.

Thoughts?

Reply:
Why is `native` discouraged? I can't really think of a strong argument for or against ENV variables in genera other than maybe it's not a good idea to overload other peoples' env variables? We then have to worry about when they may unset, if we want to retry for example.

Reply:
To venture a guess; native is discouraged because by doing so the library implementor is encouraged to look for similarities between different schedulers/job-launch paradigms, and so make better use of our existing interface.

Thinking of the risk of overloading the meaning of an existing environment variable is good advice. As of Singularity 3.x the container env var is only set within a running container, and I unset it before the container starts, but that's not to say that the Singularity team doesn't own that variable, and so may change what it does in future releases.

Reply:
To pick this back up, I am now leaning towards using `native` for image selection and specifying bindmounts. The reason is because at present the batch plugin interface does not expose the ability to set environment variables, but it does expose `script.native`.